vgen/src/main/java/io/gapi/vgen/CodeGenerator.java

```diff
@@ -40,36 +32,40 @@
  */
 public class CodeGenerator {
 
-  private final GapicLanguageProvider provider;
+  private final GapicLanguageProvider<ProtoElement> provider;
 
-  public CodeGenerator(GapicLanguageProvider provider) {
+  public CodeGenerator(GapicLanguageProvider<ProtoElement> provider) {
     this.provider = Preconditions.checkNotNull(provider);
   }
 
-  public static CodeGenerator create(ConfigProto configProto, Model model) {
-    return new CodeGenerator(GeneratorBuilderUtil.createLanguageProvider(configProto, model));
+  public static CodeGenerator create(ConfigProto configProto, TemplateProto template, Model model) {
+    return new CodeGenerator(
+        GeneratorBuilderUtil.createLanguageProvider(configProto, template, model)); // FIXME
   }
 
   /**
    * Generates code for the model. Returns a map from service interface to code for the service.
    * Returns null if generation failed.
    */
   @Nullable
-  public Map<Interface, GeneratedResult> generate(SnippetDescriptor snippetDescriptor) {
+  public Map<ProtoElement, GeneratedResult> generate(SnippetDescriptor snippetDescriptor) {
+    ProtoElementViewing<ProtoElement> viewing = provider.getViewing();
+    Iterable<ProtoElement> elts = viewing.getIterable(provider.getModel());
```

Comment:
Let's go with InputElementProvider -- it's nicely parallel with LanguageProvider once they're both in the config.